public/app/angular/panel/AngularPanelReactWrapper.tsx

```diff
@@ -0,0 +1,126 @@
+import React, { ComponentType, useEffect, useRef } from 'react';
+import { Observable, ReplaySubject } from 'rxjs';
+
+import { EventBusSrv, PanelData, PanelPlugin, PanelProps, FieldConfigSource } from '@grafana/data';
+import { AngularComponent, getAngularLoader, RefreshEvent } from '@grafana/runtime';
+import { DashboardModelCompatabilityWrapper } from 'app/features/dashboard-scene/utils/DashboardModelCompatabilityWrapper';
+import { GetDataOptions } from 'app/features/query/state/PanelQueryRunner';
+import { RenderEvent } from 'app/types/events';
+
+interface AngularScopeProps {
+  panel: PanelModelCompatabilityWrapper;
+  dashboard: DashboardModelCompatabilityWrapper;
+  queryRunner: FakeQueryRunner;
+  size: {
+    height: number;
+    width: number;
+  };
+}
+
+export function getAngularPanelReactWrapper(plugin: PanelPlugin): ComponentType<PanelProps> {
+  return function AngularWrapper(props: PanelProps) {
+    const divRef = useRef<HTMLDivElement>(null);
+    const angularState = useRef<AngularScopeProps | undefined>();
+    const angularComponent = useRef<AngularComponent | undefined>();
+
+    useEffect(() => {
+      if (!divRef.current) {
+        return;
+      }
+
+      const loader = getAngularLoader();
+      const template = '<plugin-component type="panel" class="panel-height-helper"></plugin-component>';
+      const queryRunner = new FakeQueryRunner();
+      const fakePanel = new PanelModelCompatabilityWrapper(plugin, props, queryRunner);
+
+      angularState.current = {
+        // @ts-ignore
+        panel: fakePanel,
+        // @ts-ignore
+        dashboard: new DashboardModelCompatabilityWrapper(),
+        size: { width: props.width, height: props.height },
+        queryRunner: queryRunner,
+      };
+
+      angularComponent.current = loader.load(divRef.current, angularState.current, template);
+
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+
+    // Re-render angular panel when dimensions change
+    useEffect(() => {
+      if (!angularComponent.current) {
+        return;
+      }
+
+      angularState.current!.size.height = props.height;
+      angularState.current!.size.width = props.width;
+      angularState.current!.panel.events.publish(new RenderEvent());
+    }, [props.width, props.height]);
+
+    // Pass new data to angular panel
+    useEffect(() => {
+      if (!angularState.current?.panel) {
+        return;
+      }
+
+      angularState.current.queryRunner.forwardNewData(props.data);
```

Comment:
I may be missing something here, but how come new data is consumed by the angular panel? Can't see this being done via the query runner.

Reply:
@dprokop https://github.com/grafana/grafana/blob/main/public/app/angular/panel/metrics_panel_ctrl.ts#L57

Reply:
ah, right!